// firebase-tools/lib/emulator/storage/persistence.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Persistence = void 0;
const fs_1 = require("fs");
const rimraf = require("rimraf");
const fs = require("fs");
const path = require("path");
class Persistence {
    constructor(dirPath) {
        this.reset(dirPath);
    }
    reset(dirPath) {
        this._dirPath = dirPath;
        (0, fs_1.mkdirSync)(dirPath, {
            recursive: true,
        });
    }
    get dirPath() {
        return this._dirPath;
    }
    appendBytes(fileName, bytes) {
        const filepath = this.getDiskPath(fileName);
        let fd;
        try {
            fs.appendFileSync(filepath, bytes);
            return filepath;
        }
        finally {
            if (fd) {
                (0, fs_1.closeSync)(fd);
            }
        }
    }
    readBytes(fileName, size, fileOffset) {
        let fd;
        try {
            fd = (0, fs_1.openSync)(this.getDiskPath(fileName), "r");
            const buf = Buffer.alloc(size);
            const offset = fileOffset && fileOffset > 0 ? fileOffset : 0;
            (0, fs_1.readSync)(fd, buf, 0, size, offset);
            return buf;
        }
        finally {
            if (fd) {
                (0, fs_1.closeSync)(fd);
            }
        }
    }
    deleteFile(fileName, failSilently = false) {
        try {
            (0, fs_1.unlinkSync)(this.getDiskPath(fileName));
        }
        catch (err) {
            if (!failSilently) {
                throw err;
            }
        }
    }
    deleteAll() {
        return new Promise((resolve, reject) => {
            rimraf(this._dirPath, (err) => {
                if (err) {
                    reject(err);
                }
                else {
                    resolve();
                }
            });
        });
    }
    renameFile(oldName, newName) {
        (0, fs_1.renameSync)(this.getDiskPath(oldName), this.getDiskPath(newName));
    }
    getDiskPath(fileName) {
        return path.join(this._dirPath, encodeURIComponent(fileName));
    }
}
exports.Persistence = Persistence;
